Feature/product

hyun132 opened the pull request:
- Product에 relatedProducts추가
- GetProductUseCase 수정
    - 상품명 유효성 체크
    - 연관상품 가져오는 로직 추가
    - InvalidDataException 생성 및 에러 타입 정의
- Error, DataError 생성 및 에러 타입 정의
- ResultEntity 변경 및 이를 참조하는 부분 중 상품 관련된 파일 수정.

hyun132 committed:
feature[#30] : Product에 relatedProducts추가. ProductViewModel, ProductScreen에서 연관상품 따로 호출하는 부분 제거

연관상품을 Product클래스에 추가하여 연관상품 별도로 호출 및 관리하는 로직 제거

hyun132 committed:
feature[#30] : GetProductUseCase에 상품의 연관상품 호출하는 부분 추가, 상품명 유효성 확인하는 로직 추가, InvalidDataException 및 DataError 생성.

- GetProductUseCase 상품의 연관상품 가져오는 로직 추가.
- 데이터 유효하지 않음을 표현하는 InvalidDataException 생성.
- 예외 상태 나타내는 DataError객체 생성 및 상태값 정의
- FireStoreNoDataException를 InvalidDataException로 변경

hyun132 committed:
feature[#30] : 기존 ResultEntity를 새로 정의한 Error타입을 사용하도록 수정.

변경한 ResultEntity에 맞게 handler및 Extensions 수정

hyun132 committed:
feature[#30] : 새로 정의한 ResultEntity에 맞게 Product관련 핸들러 객체 수정

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/model/mapper/ProductMapper.kt:
```diff
@@ -23,7 +23,8 @@ class ProductMapper {
             dto.masterProductCode ?: "",
             dto.masterProductName ?: "",
-            dto.dealProducts?:emptyList()
+            dto.dealProducts?:emptyList(),
```
?: 연산자 사이에 공백을 추가해주세요. :)
코틀린 Lint 룰을 추가해보는 것도 좋을 것 같아요.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/product/ProductViewModel.kt:
```diff
@@ -84,8 +57,7 @@ class ProductViewModel @Inject constructor(
         val isLoading: ProductState = ProductState.LOADING,
-        val product: Product = dummyProduct,
-        val relatedProducts: List<ProductPreview> = emptyList()
+        val product: Product = dummyProduct
```
ViewModel에서 dummyProduct를 디폴트로 주는 이유가 있을까요?

hyun132 replied:
데이터를 가져오기 전이나 데이터 로드를 실패했을 때 화면에 기본값이라도 보여주는게 나을까 싶어서 더미데이터를 사용했습니다!

f-lab-dean replied:
데이터 가져오기 전에는 로딩 표시, 실패했을 때는 에러 핸들링을 추가하면 어떨까요?

hyun132 replied:
넵 로딩이랑 에러 핸들링 추가해보겠습니다! 그리고 초기값으로 더미데이터를 주지 않는게 좋을까요?

f-lab-dean replied:
더미데이터를 주면 화면이 어떻게 표현될까요?
어떤 화면이 나타나는지에 따라 다를 것 같습니다.
무의미한 데이터가 나오거나, 아무것도 보이지 않는다면 큰 의미가 없을 것 같아요.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/model/common/Exceptions.kt:
```diff
@@ -1,3 +1,5 @@
\ No newline at end of file
+
+class FireStoreNoDataException(override val message: String?):Exception(message)
+class InvalidDataException(val type : DataError = DataError.VALIDATION.UNKNOWN):Exception()
```
: 사이에 공백을 확인해주세요.

hyun132 replied:
넵 수정하겠습니다!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/usecases/product/GetProductUseCase.kt:
```diff
@@ -1,17 +1,52 @@
 ) {
-    suspend fun invoke(id: String): Flow<ResultEntity<Product>> =
-        repository.getProduct(id).mapToResultEntity(errorHandler)
+    private val RELATED_PRODUCT_SIZE = 5
```
상수는 companion object에 const로 선언하는게 어떤가요?

hyun132 replied:
수정하겠습니다!!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/usecases/product/GetProductUseCase.kt:
```diff
@@ -1,17 +1,52 @@
+        }.mapToResultEntity(errorHandler)
+
+    private fun checkProductValidation(product: Product) {
+        if (product.name.isEmpty() || product.name.isBlank()) throw InvalidDataException(DataError.VALIDATION.MISSING_VALUE)
```
throw를 하지 않는 방법은 없을까요?

hyun132 replied:
throw를 하지 않고 Error값을 흘려보내게 수정해보겠습니다!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/usecases/product/GetProductUseCase.kt:
```diff
@@ -1,17 +1,52 @@
+    }
+
+    private suspend fun searchRelatedProductIfKeywordIsValid(product: Product): Flow<Product> {
+        val lastRightBracketIndex = product.name.lastIndexOf('[')
```
정규표현식을 사용하는 방법은 어떤가요?

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/usecases/base/BaseUseCase.kt:
```diff
@@ -1,5 +1,5 @@
 
-abstract class BaseUseCase<in P, R> {
-    abstract suspend operator fun invoke(parameters: P): R
+abstract class BaseUseCase<R> {
```
interface가 아닌 abstract class를 사용한 이유는 무엇인가요?

hyun132 replied:
처음 객체 생성할 때 파라미터를 필드로 가지고 있게 하려고 했었던 걸로 기억합니다! 현재는 사용하고 있지 않은 파일이라 삭제하겠습니다

hyun132 committed:
feature[#30] : [피드백 반영] 코드 컨벤션에 맞게 공백 수정

hyun132 committed:
feature[#30] : [피드백 반영] 상품명이 빈 값일 때 예외 throw대신 Error타입 내보내도록 수정

hyun132 committed:
feature[#30] : [피드백 반영] 상수 companion object에 const val로 수정

hyun132 committed:
feature[#30] : [피드백 반영] 사용하지 않는 파일 제거

hyun132 committed:
feature[#30] : [피드백 반영] : 로딩시 프로스레스바, 에러 발생시 다이얼로그 띄우기

데이터 로딩시 CircularProgressIndicator를 보여주고, 에러 발생 시 오류발생을 알리는 다이얼로그를 띄우고 이전화면으로 돌아가도록 수정

f-lab-dean commented on app/src/main/java/com/example/commerceapp/domain/model/common/Error.kt:
```diff
@@ -0,0 +1,3 @@
+package com.example.commerceapp.domain.model.common
+
+sealed interface Error
```
현재 하위 타입이 없으므로 일반 interface로 선언해도 될 것 같습니다.
야그니(YAGNI) 원칙을 고려해보세요.
> https://ko.wikipedia.org/wiki/YAGNI

f-lab-dean commented on app/src/main/java/com/example/commerceapp/data/remote/ProductRepositoryImpl.kt:
```diff
@@ -58,7 +60,7 @@ class ProductRepositoryImpl @Inject constructor(
             .mapNotNull { snapshot ->  // null이면 무시
                 snapshot.firstOrNull()?.toObject(ProductDto::class.java)
-                    ?.let { productMapper.mapToProduct(it) } ?: throw FireStoreNoDataException("데이터가 없습니다.")
+                    ?.let { productMapper.mapToProduct(it) } ?: throw InvalidDataException(DataError.VALIDATION.EMPTY)
```
여기도 UseCase와 동일한 맥락으로 throw 하지 않는건 어떤지 고려해봐도 좋을 것 같아요.

f-lab-dean reviewed (approved):
나머지 코멘트는 현재 PR 또는 다른 PR에서 반영을 고민해주세요.
LGTM!

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/product/ProductScreen.kt:
```diff
@@ -155,6 +163,48 @@ fun ProductScreen(
+    }
+}
+
+suspend fun loadProgress(updateProgress: (Float) -> Unit) {
```
사용되는 곳이 있는 코드인가요?

hyun132 replied:
앗 아니요 수정하겠습니다

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/product/ProductScreen.kt:
```diff
@@ -70,83 +72,89 @@ fun ProductScreen(
+                            modifier = Modifier.padding(16.dp),
+                        )
+                        LazyRow(
+                            modifier = Modifier.padding(start = 8.dp),
```
스크롤 가능한 컨텐츠들(LazyColumn, LazyRow)를 다룰 때, Modifier의 padding대신 contentPadding을 활용할 수 있습니다.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/product/ProductScreen.kt:
```diff
@@ -70,83 +72,89 @@ fun ProductScreen(
+                item(uiState.product.no) {
+                    HorizontalDivider(thickness = 8.dp, color = gray1)
+                    Column {
+                        Text(
```
각각의 아이템을 가독성과 재사용성을 위해 별도의 Composable로 추출하는 것을 권장합니다.

f-lab-dean commented on app/src/main/java/com/example/commerceapp/ui/product/ProductScreen.kt:
```diff
@@ -70,83 +72,89 @@ fun ProductScreen(
+            topBar = { TopAppbar(navController, uiState.product.name) },
+            bottomBar = {
+                Row(
                     modifier = Modifier
```
가독성과 재사용성을 위해 별도의 Composable로 추출하는 것을 권장합니다.

hyun132 replied:
넵 수정하겠습니다!

hyun132 committed:
사용하지 않는 프로그레스바 삭제

hyun132 committed:
하나의 scaffold안에 구성되어있던 컴포저블을 별도의 컴포저블 함수로 분리

hyun132 commented:
잔여 피드백 및 수정사항은 별도 pr로 진행하도록 하겠습니다
